File: conference-js/src/ConferencePresenter.js
import { State } from "./State";

export class ConferencePresenter {
  constructor(viewModel) {
    this.viewModel = viewModel;
  }

  onStateChanged(state) {
    switch (state.name) {
      case State.LOADING:
        this.viewModel.showLoading = true;
        this.viewModel.showError = false;
        this.viewModel.showConference = false;
        break;
      case State.LOADED:
        this.viewModel.showLoading = false;
        this.viewModel.showError = false;
        this.viewModel.showConference = true;
        this.viewModel.conferenceName = state.value.name;
        break;
      case State.ERROR:
        this.viewModel.showLoading = false;
        this.viewModel.showError = true;
        this.viewModel.errorMessage = state.error.message;
        this.viewModel.showConference = false;
        break;
      case State.IDLE:
      default:
        this.viewModel.showLoading = false;
        this.viewModel.showError = false;
        this.viewModel.showConference = false;
        break;
    }
    this.viewModel.notifyObservers();
  }
}

File: conference-android/app/src/main/java/com/droidcon/conference/ConferencePresenter.kt
package com.droidcon.conference

import com.droidcon.gateway.GatewayError
import com.droidcon.state.State
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

class ConferencePresenter(state: Flow<State<Conference, GatewayError>>) {
    val viewModel = state.map { state ->
        val viewModel = ConferenceViewModel()
        when (state.name) {
            State.Name.IDLE -> {
                viewModel.showLoading = false
                viewModel.showError = false
                viewModel.hideName = true
                viewModel.showRetry = false
            }
            State.Name.LOADING -> {
                viewModel.showLoading = true
                viewModel.showError = false
                viewModel.hideName = true
                viewModel.showRetry = false
            }
            State.Name.LOADED -> {
                val conference = state.value!!
                viewModel.showLoading = false
                viewModel.showError = false
                viewModel.hideName = false
                viewModel.showRetry = false
                viewModel.name = conference.name
            }
            State.Name.ERROR -> {
                viewModel.showLoading = false
                viewModel.showError = true
                viewModel.hideName = true
                viewModel.showRetry = true
            }
        }
        viewModel
    }
}
